### PR TITLE
limit maximal desired horizontal acceleration

### PR DESCRIPTION
**Pull request by tumbili**

### CODE REVIEW

**Comment by tumbili** [comment]
Tested in SITL, feel really nice. The response to stick inputs is much smoother.

**Comment by tumbili** [comment]
no_acc_lim

**Comment by tumbili** [comment]
acc_lim

**Comment by LorenzMeier** [comment]
Two things:

  * The prev vel SP should be checked with PX4_ISFINITE on each of its elements
  * Does it maybe need to be reset with the position SP on mode switches here? https://github.com/PX4/Firmware/blob/mc_pos_max_acc/src/modules/mc_pos_control/mc_pos_control_main.cpp#L652

**Comment by LorenzMeier** [comment]
After further inspection I'm 100% sure it needs those resets, as you will get weird results else if you switch between auto and manual. The _last_vel_sp should be reset to the current velocity on these occasions if the current position / velocity is valid.

**Comment by LorenzMeier** [comment]
@tumbili Rebased, added missing checks.

**Comment by LorenzMeier** [comment]
@AndreasAntener @sjwilks Could you test this maybe today? It will be needed for our new joint project (which is part of the reason we did this now).

Things to test:

  * Transition manual -> pos ctrl -> manual -> posctrl *AT SPEED*
  * Transition pos ctrl -> auto -> pos ctrl -> auto *AT SPEED*
  * Transition manual -> auto -> manual -> auto *AT SPEED*
  * Flying in pos ctrl with aggressive stick inputs. Try in particular flying at speed and then letting go of the stick.
  * RTL: Once you hit RTL, you should see a smooth action.

The velocity doesn't need to be super high, but something like 3-8 m/s. You should see at no point a rapid twitch but only gradual deceleration and acceleration.

**Comment by LorenzMeier** [comment]
Needs inspection of the attitude setpoint. We need to make sure it does something sensible on switching as well there.

We might need to limit the body frame acceleration as well, so that thrust steps are impossible. That will almost certainly damp mode switch twitches to a convenient level:
https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_main.cpp#L815

**Comment by LorenzMeier** [comment]
We might in fact want to limit the jerk using limiters here: https://github.com/PX4/Firmware/blob/master/src/modules/mc_att_control/mc_att_control_main.cpp#L860-L863

**Comment by AndreasAntener** [comment]
Here's the log where we tried to increase the acceleration parameter: http://logs.uaventure.com/view/HxW6SA3opUXtYPY2Evw7jR
And an RTL: http://logs.uaventure.com/view/4oJn2ByFLpQufgwJJkgh4i

We're not sure if this was an effect of the changed param, but when we switched from manual to auto (went into loiter), it started to swing forth and back. Same for RTL

**Comment by LorenzMeier** [comment]
Yeah, clearly visible step in attitude setpoint. We will need to get rid of that.

screen shot 2015-12-07 at 16 21 13

**Comment by LorenzMeier** [comment]
The last commit should fix this initialization issue. Needs a SITL and a real test flight.

**Comment by LorenzMeier** [comment]
@AndreasAntener Can you test the last state in SITL and continue to outdoors if it passes SITL? It would be good if you drive this. We need to resolve this for the scope of your current project as it looks horrible without.

**Comment by AndreasAntener** [comment]
Tried jmavsim sitl, in posctl it twiches every time the stick input changes. As transition from manual/altctl to posctl it steps on the breaks and "swings back"

**Comment by LorenzMeier** [comment]
@AndreasAntener Hrmpf. Thanks! I'll give it another go tomorrow morning.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -1440,6 +1497,36 @@ MulticopterPositionControl::task_main()
...
+
+					/* this is jerk times dt
+					 * we call it jerk so its clear what the intention is -
+					 * to limit the jerk, not the total acceleration.
+					 */
+					jerk_hor(0) = thrust_diff(0);
+					jerk_hor(1) = thrust_diff(1);
+
+					/* because we compare with jerk times dt, we multiply here as well */
+					float max_hor_jerk = (_params.acc_hor_max / ONE_G) * dt;
```
Comment (tumbili) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: @LorenzMeier If I understand the line above correctly then you are limiting the system to only increase acceleration by acc_hor_max / G in one second. So after one second the vehicle will have a velocity of only 0.5 m/s.
I think you should be removing the G in the denominator.

Comment (LorenzMeier) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: No. I'm bringing the acceleration from m/s^2 to the normalised unit of throttle.

**Comment by LorenzMeier** [comment]
@AndreasAntener This behaves in SITL (which doesn't have great tuning still) relatively well. Can you please give this another test on your quad? I rebased the branch, so you will need to delete your local one and check out freshly.

**Comment by weiweikong** [comment]
Could you explain the mathmatical detail about the function of `reset_pos_sp` ?

**Comment by LorenzMeier** [comment]
There is nothing mathematical about it. The controller tries to hold the last valid position, and that depends on the flight mode - that is all there is to it.

**Comment by weiweikong** [comment]
Thank you for your quick reply. I just read the comment section several days before. Thanks. 
Below is the original one,
```
//+ (_vel(0) - PX4_R(_att_sp_msg.data().R_body, 0, 2) * _att_sp_msg.data().thrust / _params.vel_p(0)
				// - _params.vel_ff(0) * _sp_move_rate(0)) / _params.pos_p(0);
```

**Comment by AndreasAntener** [comment]
@LorenzMeier 
Just did a few flights:
http://logs.uaventure.com/view/cjgTZMVShGR7GHPPqGeXUN (here you can see transitions)
http://logs.uaventure.com/view/E5yd6ixnxatJ285bmHCX4B (getting a feel for the controls)
http://logs.uaventure.com/view/BM3LVfoVAEWma7MLDsgKaX (and doubled the max_acc)

It flies pretty nice already :). Haven't tried RTL and auto yet.

Issues:
- yaw in posctl doesn't work
- transition from altctl to posctl makes it break first, the other way around is fine

**Comment by LorenzMeier** [comment]
@AndreasAntener I should have fixed both in the last commit. Please re-try in SITL (no gamepad here ATM), if it behaves (it might still move oddly, but should have yaw) then go ahead and fly it.

Do you have a Phantom? We target the same acceleration behaviour as it for the default accel param.

**Comment by AndreasAntener** [comment]
Yaw is working again. The transitions with speed into posctl result in a big spiraling motion and is practically uncontrollable for a while (until it settles).

**Comment by LorenzMeier** [comment]
Thats in SITL, Right? Can you put the log of that into log muncher? I would not be surprised at all if it's correct for the real system.

**Comment by tumbili** [comment]
@LorenzMeier @AndreasAntener There is a problem in SITL with master. I checked out a commit of last week and cherry-picked all the changes of this branch. I don't see any problems in SITL.

**Comment by LorenzMeier** [comment]
Can you give me the commit you checked out?

**Comment by tumbili** [comment]
c4ed598

**Comment by tumbili** [comment]
@LorenzMeier I noticed immediately that something was not right, because when I set throttle to zero the quad only descends very slowly in manual mode (throttle doesn't want to go to zero)

**Comment by AndreasAntener** [comment]
@tumbili I noticed that throttle issue as well, but I just flew it in real and I have the same issue as in sitl. Let's see if I can get the log out of the safe.

**Comment by AndreasAntener** [comment]
To clarify: I don't have the sitl throttle issue in real. But the spiraling effect when coming in fast from altctl to posctl.

**Comment by LorenzMeier** [comment]
Look at this:
https://github.com/PX4/Firmware/commit/988ff591030ba16484c256b0e05bb94d5f116136

I just cherry-picked it here as well.

**Comment by LorenzMeier** [comment]
We might need to back out these changes as well:
```
git diff c4ed598 posix-configs/SITL/init/rcS_jmavsim_iris
```

**Comment by LorenzMeier** [comment]
@tumbili @AndreasAntener Rebased on master, which has the IO stack smashing fix.

**Comment by LorenzMeier** [comment]
@tumbili Filed a new issue for yaw handling which might be related:
https://github.com/PX4/Firmware/issues/3359

**Comment by AndreasAntener** [comment]
So here's a new SITL log: http://logs.uaventure.com/view/Rbr48eGkosLwLofsJkSmC9

The issues we're still facing are:
- when transitioning from altctl to posctl at speed, it jumps to full speed in the same direction and ends in a very large spiral. it's not controllable anymore and you need to switch out of posctl. see the last transition in the log.
- in posctl, when the stick input changes, it gives small but noticeable twitch before the acceleration starts (also in real). doesn't matter if it's already moving or if the sticks were centered before

**Comment by LorenzMeier** [comment]
@AndreasAntener @tumbili I would like to get this in as soon as possible, this is the last bit missing for the next release. Is there anything holding us back?

Can you fly this in mission tomorrow?

**Comment by LorenzMeier** [comment]
@AndreasAntener Let me know if this flight-tested.

**Comment by AndreasAntener** [comment]
Good to go: http://logs.uaventure.com/view/LQ7UYHxfLpoq6RvzLE5EGF

**Comment by LorenzMeier** [comment]
Ok, applied to master, thanks!

**Pull request closed without merging**